Remove any views

digitalheir opened the pull request:
Rebased afa5a2bfe5fd551a0d5d951328c27e75884b3906 to fix demo. I've reverted `standardButtonBuilder` to its old signature (returning `AnyView` and made it deprecated in favor of strongly-typed `defaultButtonBuilder`. Sorry about breaking everything...

Picking up the thread from https://github.com/KeyboardKit/KeyboardKit/pull/357#issuecomment-986607226:
> In this case, I don't think it brings much value, since it basically just creates a context menu that iterates over the items, which basically just is what the context menu does.

I remember my use-case now. I wanted to have two buttons: one for subtype (alphabet/language) and one for keyboard layout (QWERTY/AZERTY), flattening it all to a single locale picker would make a combinatorial explosion.

It's because I wanted to make a math-specific keyboard where the subtypes are different fields of mathematics and common symbols for that field appear as the letters' alt keys, but users also want to differentiate locales. (And I don't want to create a separate keyboard plugin for each subtype.)

digitalheir commented:
Also:

> I also wanted to ask you about the builder signature, where you've added a layout parameter. I'm curious, in what way do you use the layout in your keyboard?

I don't remember why I added this param, maybe it is unnecessary. I had suspended development on my app until this PR settled so I'm kind of out of it now, I'll see if I run into problems.

danielsaidi commented:
> I don't remember why I added this param, maybe it is unnecessary. I had suspended development on my app until this PR settled so I'm kind of out of it now, I'll see if I run into problems.

The signature should really just need the item contextual values. You could just use the layout that you pass into the SystemKeyboard in the builder block right away, so I see no need in having it as a parameter. :)

danielsaidi commented:
Btw, I think that is someone is using the old, static `standardButtonBuilder`, this commit will break that, right? I mean, since it reames it to `defaultButtonBuilder`.

digitalheir commented:
No, [look](https://github.com/KeyboardKit/KeyboardKit/pull/363/files#diff-c8900d1696d89f53bbe81b2eb58cb7877f67481fc2d92b4ee1d929c9c47241f7R244), `standardButtonBuilder` is still available. 

I've split the function into `standardButtonBuilder`, which still returns `AnyView` and added `defaultButtonBuilder`, which does not type-erase. (Changed the name to reduce chance of collision when Swift does resolution.)

digitalheir commented:
If Swift is not smart enough, I guess we should rename `SystemKeyboard` to `SystemKeyboardTyped` (or something), and type-alias `SystemKeyboard` to `SystemKeyboardTyped<AnyView>`...

I think it's not so pretty, but it may be more solid for being backwards compatible.

danielsaidi commented:
Great job, many, many thanks! I am going through the changes now. I will stick to the term `buttonBuilder` since `rowItem` sounds like the item to use for an entire row. However, with your changes, I think we'll get an initializer that has a `buttonBuilder` and one that has a `buttonContentBuilder`, which will make things a LOT easier to understand than in previous versions.

danielsaidi commented:
I just realized that making `SystemKeyboard` generic is a breaking change in itself, since the static `standardButtonBuilder` is now only available if the content type is `AnyView`.

danielsaidi commented:
Also I have to replace the global default and standard functions with proper extensions.

danielsaidi commented:
@digitalheir I have pushed a [WIP] commit where I struggle with replacing the global `standardSystemKeyboard` functions with real initializers. 

Please have a look as well. This is what is so hard with generics, and why I eventually resorted to `AnyView`. I would love to get this generic approach to work, but it must be through initializers and not global functions.

danielsaidi commented:
@digitalheir I solved it! The standard and custom button view initializers are fully generic, but the custom button content initializer must use `AnyView` to abstract the content view. However, that is just an implementation detail and doesn't require us to wrap the views in an `AnyView`, so I think that it's actually pretty nice.

I have cleaned up the code and the previews and now also have previews for the standard keyboard and the buttons content replacing keyboard as well.

danielsaidi commented:
I have merged this into `master` now 🎉 

Thank you so much for this initiative. I have created more tickets for the remaining AnyView usages, so hopefully we can get rid of them as well.

digitalheir commented:
Happy to work on it myself, I'll take a look tomorrow. (And promise I'll be more careful about breaking changes...)

digitalheir commented:
> the custom button content initializer must use AnyView to abstract the content view

Do you mean that the `View` type-erases the internal hierarchy? I'm not in favour of this. My initial motivation was actually more inclined to SwiftUI performance (maintaining structural identity) than programming ergonomics...

danielsaidi commented:
Take a look at the initializer. Perhaps you can figure out how to define the generic so that we don’t need the AnyView?
